### PR TITLE
Improved `DenseLQDynamicModel` construction to require less memory and build faster

### PR DESCRIPTION
**Pull request by dlcole3**
Reformulated source code to build the Hessian and Jacobian a block at a time for the dense model. Previously, all block matrices were built, stored, and then multiplied together, but this was very inefficient as many of the multiplications were redundant. Instead, each block within the block matrices were multiplied together only when needed within for loops. 

This PR includes:
- redefining `DenseLQDynamicBlocks` to only save the `A` block and the `B` block. The latter block only contains the first column of the actual block `B` matrix, and it does not contain the first rows of zeros. This also removed any sparse matrices, which were inefficient within `LinearAlgebra.mul!`
- redefining `_build_block_matrices` to match `DenseLQDynamicBlocks`.
- redefining `_build_H_blocks` 
- redefining `_build_G_blocks`. 
- defining a new function, `_build_dense_J1` to form the Jacobian from the `G` blocks
- updating `build_dense_lq_dynamic_model` to be consistent with these changes
- updating `get_s` and `get_u` for working with the new block `B` matrix

This PR does not include updating the sparse formulation (which is also somewhat inefficient) or redefining the internal functions to be non-allocating. These changes are still needed but will be met with other PRs.

### CODE REVIEW

File: src/DynamicNLPModels.jl [hunk 1]
```diff
@@ -115,9 +117,9 @@ function LQDynamicData(
     N;
 
     Qf::M = Q, 
-    S::M  = (similar(Q, size(Q, 1), size(R, 1)) .= 0),
-    E::M  = (similar(Q, 0, length(s0)) .= 0),
-    F::M  = (similar(Q, 0, size(R, 1)) .= 0),
+    S::M  = (similar(Q, size(Q, 1), size(R, 1)) .= 0.0),
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: This syntax is too heavy and hard to parse. Prefer: 
```julia
S = similar(Q, n, m) ; fill!(S, 0,0)
```

File: src/DynamicNLPModels.jl [hunk 2]
```diff
@@ -865,23 +876,23 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
-        ucon[(length(ucon1) + length(ucon2) + 1):(length(ucon1) + length(ucon2) + length(ucon3))] .= ucon3
+        lcon[(length(dl) + length(lcon2) + 1):(length(dl) + length(lcon2) + length(lcon3))] .= lcon3
+        ucon[(length(du) + length(ucon2) + 1):(length(du) + length(ucon2) + length(ucon3))] .= ucon3
     end
 
 
     nvar = nu * N
     nnzj = size(J, 1) * size(J, 2)
-    nnzh = sum(LinearAlgebra.LowerTriangular(H) .!= 0)
+    nnzh = floor(Int, size(H, 1) * (size(H,1) + 1) / 2)
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: For readability, prefer:
```julia
nh = size(H, 1)
nnzh = div(nh * (nh + 1), 2)
```

File: src/DynamicNLPModels.jl [hunk 3]
```diff
@@ -115,9 +117,9 @@ function LQDynamicData(
     N;
 
     Qf::M = Q, 
-    S::M  = (similar(Q, size(Q, 1), size(R, 1)) .= 0),
-    E::M  = (similar(Q, 0, length(s0)) .= 0),
-    F::M  = (similar(Q, 0, size(R, 1)) .= 0),
+    S::M  = (similar(Q, size(Q, 1), size(R, 1)) .= 0.0),
+    E::M  = (similar(Q, 0, length(s0)) .= 0.0),
+    F::M  = (similar(Q, 0, size(R, 1)) .= 0.0),
```
Comment (sshin23) [inline src/DynamicNLPModels.jl]: zero(T)

File: src/DynamicNLPModels.jl [hunk 4]
```diff
@@ -203,53 +205,30 @@ struct SparseLQDynamicModel{T, V, M1, M2, M3, MK} <:  AbstractDynamicModel{T,V}
...
-    K::M2
-    E::M1
-    F::M1
-    gl::M1
-    gu::M1
+struct DenseLQDynamicBlocks{T, M}
+    A::M
+    B::M
+
 end
```
Comment (sshin23) [inline src/DynamicNLPModels.jl]: Maybe good to add a docstring of explaining what are these. Or we can even point to the reference

File: src/DynamicNLPModels.jl [hunk 5]
```diff
@@ -203,53 +205,30 @@ struct SparseLQDynamicModel{T, V, M1, M2, M3, MK} <:  AbstractDynamicModel{T,V}
...
 
 function DenseLQDynamicBlocks(
```
Comment (sshin23) [inline src/DynamicNLPModels.jl]: is this constructor even necessary? can't we just use the default constructor?

File: src/DynamicNLPModels.jl [hunk 6]
```diff
@@ -292,9 +271,9 @@ function SparseLQDynamicModel(
     R::M,
     N;
     Qf::M = Q, 
-    S::M  = (similar(Q, size(Q, 1), size(R, 1)) .= 0),
-    E::M  = (similar(Q, 0, length(s0)) .= 0),
-    F::M  = (similar(Q, 0, size(R, 1)) .= 0),
+    S::M  = (similar(Q, size(Q, 1), size(R, 1)) .= 0.0),
```
Comment (sshin23) [inline src/DynamicNLPModels.jl]: zero(T)

File: src/DynamicNLPModels.jl [hunk 7]
```diff
@@ -609,30 +588,33 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
 
-    H_blocks = _build_H_blocks(block_Q, block_R, block_A, block_B, block_S, block_K, s0, N, K)
+
+    H_blocks = _build_H_blocks(Q, R, block_A, block_B, S, s0, N, Qf, K)
 
     H  = H_blocks.H
     c0 = H_blocks.c0
 
-    G_blocks = _build_G_blocks(block_A, block_B, block_E, block_F, block_K, block_gl, block_gu, s0, N)
+    G  = similar(Q, nc * N, nu); fill!(G, 0.0)
```
Comment (sshin23) [inline src/DynamicNLPModels.jl]: `zero(T)`

File: src/DynamicNLPModels.jl [hunk 8]
```diff
@@ -609,30 +588,33 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
+    J1 = similar(Q, nc * N, nu * N); fill!(J1, 0.0)
```
Comment (sshin23) [inline src/DynamicNLPModels.jl]: ok, this is the guy we want to eliminate later

File: src/DynamicNLPModels.jl (hunk 2 above)
Comment (sshin23) [inline src/DynamicNLPModels.jl]: `div( ... , 2)`

**Review by sshin23 — APPROVED** [review]
Looks good, overall @dlcole3. Have some minor comments

File: src/DynamicNLPModels.jl (hunk 7 above)
Comment (frapac) [inline src/DynamicNLPModels.jl]: This is the dense Jacobian, right?

File: src/DynamicNLPModels.jl [hunk 9]
```diff
@@ -609,30 +588,33 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
+    dl = similar(s0, nc * N); fill!(dl, 0.0)
+    du = similar(s0, nc * N); fill!(du, 0.0)
+
+    for i in 1:N
+        dl[(1 + nc * (i - 1)): nc * i] = gl
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: use the broadcast operator here: `dl[...] .= gl`

File: src/DynamicNLPModels.jl [hunk 10]
```diff
@@ -646,17 +628,22 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
-    As0_bounds = similar(s0, num_real_bounds * N); fill!(As0_bounds, 0)
+    J2         = similar(Q, num_real_bounds * N, nu * N); fill!(J2, 0.0)
+    As0_bounds = similar(s0, num_real_bounds * N); fill!(As0_bounds, 0.0)
 
     if num_real_bounds == length(sl)
-        J2        .= block_B[(1 + ns):end,:]
-        As0_bounds .= As0[(1 + ns):end,1]
+        As0_bounds .= As0[(1 + ns):end]
+        for i in 1:N
+            J2[(1 + (i - 1) * ns):end, (1 + nu * (i - 1)):(nu * i)] .= block_B[1:(ns * (N - i + 1)),:]
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: I would advise replacing `end` by the explicit indexing, to robustify the code

File: src/DynamicNLPModels.jl (hunk 9 above)
Comment (frapac) [reply src/DynamicNLPModels.jl]: or more directly: 
```julia
dl = repeat(gl, N)
```

File: src/DynamicNLPModels.jl [hunk 11]
```diff
@@ -749,30 +736,33 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
     gl = dnlp.gl
     gu = dnlp.gu
 
+    nc = size(E, 1)
 
-    dense_blocks = _build_block_matrices(s0, Q, R, A, B, E, F, N, gu, gl, K; Qf = Qf, S = S)
+    dense_blocks = _build_block_matrices(A, B, K, N)
```
Comment (frapac) [inline src/DynamicNLPModels.jl:704]: hehe, just a remark: here `N` is an integer, but uses the same notation in capital as the other matrices

File: src/DynamicNLPModels.jl [hunk 12]
```diff
@@ -749,30 +736,33 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
+    J1 = similar(Q, nc * N, nu * N); fill!(J1, 0.0)
+    dl = similar(s0, nc * N); fill!(dl, 0.0)
+    du = similar(s0, nc * N); fill!(du, 0.0)
 
-    J1    = G_blocks.J
-    lcon1 = G_blocks.lcon
-    ucon1 = G_blocks.ucon
-    As0   = G_blocks.As0
+    for i in 1:N
+        dl[(1 + nc * (i - 1)): nc * i] = gl
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: Use the broadcast operator here, or `repeat`

File: src/DynamicNLPModels.jl [hunk 13]
```diff
@@ -800,30 +795,46 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
+        if i == 1
+            KB = I_mat
+        else
+            B_sub_block = block_B[(1 + ns * (i - 2)):ns * (i - 1), :]
+            LinearAlgebra.mul!(KB, K, B_sub_block)
+        end
+        
+        KBI[(1 + nu * (i - 1)):(nu * i),:] = KB
+        LinearAlgebra.mul!(KAs0_temp, K, As0[(1 + ns * (i - 1)):ns * i])
+        KAs0[(1 + nu * (i - 1)):nu * i] = KAs0_temp
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: broadcast operator

File: src/DynamicNLPModels.jl [hunk 14]
```diff
@@ -800,30 +795,46 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
+    KBI  = similar(Q, nu * N, nu)
     KAs0 = similar(s0, nu * N)
+    KAs0_temp = similar(s0, nu)
+    KB   = similar(Q, nu, nu)
 
-    I_mat = similar(Q, nu * N, nu * N); fill!(I_mat, 0)
+    I_mat = similar(Q, nu, nu); fill!(I_mat, 0)
 
-    for i in 1:(nu * N)
+    for i in 1:nu
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: Or directly: `I_mat[diagind(I_mat)] .= 1.0`

File: src/DynamicNLPModels.jl [hunk 15]
```diff
@@ -800,30 +795,46 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
+            J3[(1 + (i - 1) * nu):end, (1 + nu * (i - 1)):(nu * i)] .= KBI[1:(nu * (N - i + 1)),:]
+        end
     else
         for i in 1:N
             row_range   = (1 + (i - 1) * num_real_bounds):(i * num_real_bounds)
-            J3[row_range, :] .= KBI[(1 + nu * (i - 1)):(nu * i), :][bool_vec_u, :]
             KAs0_bounds[row_range]      .= KAs0[(1 + nu * (i - 1)):(nu * i)][bool_vec_u]
+
+            for j in 1:(N - i +1)
+            J3[(1 + (i + j - 2) * num_real_bounds):((i + j - 1) * num_real_bounds), (1 + nu * (i - 1)):(nu * i)] .= KBI[(1 + (j - 1) * nu):(j * nu), :][bool_vec_u, :]
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: indentation!

File: src/DynamicNLPModels.jl [hunk 16]
```diff
@@ -913,99 +924,55 @@ end
...
     A_klast  = copy(A_k)
     A_knext  = copy(A_k)
-    AB_klast = similar(Q, size(B, 1), size(B, 2)); fill!(AB_klast, 0)
-    AB_k     = similar(Q, size(B, 1), size(B, 2)); fill!(AB_k, 0)
+    AB_klast = similar(A, size(B, 1), size(B, 2)); fill!(AB_klast, 0.0)
+    AB_k     = similar(A, size(B, 1), size(B, 2)); fill!(AB_k, 0.0)
   
     # Fill the A and B matrices
     for i in 1:(N - 1)
         if i == 1
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: I don't like this syntax. I prefer to initialize the stage `i==1` before entering in the for loop, and then iterates with `for i in 2:(N-1)`

File: src/DynamicNLPModels.jl [hunk 17]
```diff
@@ -913,99 +924,55 @@ end
...
-            block_A[(ns + 1):ns*2, :] = A_k
-            LinearAlgebra.mul!(AB_k, A_k, B)
-            for k in 1:(N-i)
-                row_range = (1 + (k + 1) * ns):((k + 2) * ns)
-                col_range = (1 + (k - 1) * nu):(k * nu)
-                block_B[row_range, col_range] = AB_k
-            end
+            block_A[(ns + 1):ns * 2, :] = A_k
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: Broadcast!

File: src/DynamicNLPModels.jl [hunk 18]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
     )
 end
 
-function _build_H_blocks(block_Q, block_R, block_A::M, block_B::M, block_S, block_K, s0, N, K::MK) where {T, M <: AbstractMatrix{T}, MK <: Nothing}
-    As0      = similar(s0, size(block_A, 1))
-    QB       = similar(block_A, size(block_Q, 1), size(block_B, 2))
-    STB      = similar(block_A, size(block_S, 2), size(block_B, 2))
-    B_Q_B    = similar(block_A, size(block_B, 2), size(block_B, 2))
+
+function _build_H_blocks(Q, R, block_A::M, block_B::M, S, s0, N, Qf, K::MK) where {T, M <: AbstractMatrix{T}, MK <: Nothing}
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: Be careful with the order of the arguments. To be consistent, it's better to put the matrices first, then the scalar arguments.

File: src/DynamicNLPModels.jl [hunk 19]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+
+    ns = size(Q, 1)
+    nu = size(R, 1)
+
+    H  = similar(block_A, nu * N, nu * N); fill!(H, 0.0)
+
+    QB  = similar(block_B); fill!(QB, 0.0)
+    QfB = similar(block_B); fill!(QfB, 0.0)
+
+    QAB  = similar(block_A, ns, nu)
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: When we use `similar`, always initiate the resulting pointer with `0`!

File: src/DynamicNLPModels.jl [hunk 20]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+    BQfB = similar(block_B, nu, nu)
+
+    STB = similar(block_B, nu, nu)
+
+    As0 = similar(block_A, ns * (N + 1))
+
+    LinearAlgebra.mul!(As0, block_A, s0)
+
+    for i in 1:N
+        B_sub_block = block_B[(1 + (i - 1) * ns):(i * ns),:]
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: I guess this lead to unneeded allocations. Please consider using a view here

File: src/DynamicNLPModels.jl [hunk 21]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+        LinearAlgebra.mul!(QAB, Q, B_sub_block)
+        LinearAlgebra.mul!(QfAB, Qf, B_sub_block)
+        
+        QB[(1 + (i - 1) * ns):(i * ns), :]  = QAB
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: Broadcast!

File: src/DynamicNLPModels.jl [hunk 22]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+            right_block = block_B[(1 + (j - 1 + i - 1) * ns):((j + i - 1)* ns),:]
+            LinearAlgebra.mul!(BQB, QAB', right_block)
+            LinearAlgebra.mul!(BQfB, QfAB', right_block)
+
+            for k in 1:(N - j - i + 1 + 1)
+                row_range = (1 + nu * (k + (j - 1) - 1)):(nu * (k + (j - 1)))
+                col_range = (1 + nu * (k - 1)):(nu * k)
+
+                if k == N - j - i + 1 + 1
+                    view(H, row_range, col_range) .+= BQfB
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: genuine question: do we need a view here?

File: src/DynamicNLPModels.jl [hunk 23]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+    As0QB = similar(QB, nu)
+    QAs0  = similar(QB, ns)
+    As0S  = similar(block_B, nu)
+
+
+    for i in 1:N
+        fill!(QB_block_vec, 0.0)
+        rows_QB           = 1:(ns * (N - i)) 
+        rows_QfB          = (1 + ns * (N - i)):(ns * (N - i + 1))
+        QB_block_vec[(1 + ns * i):(ns * N), :] = QB[rows_QB, :]
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: broadcast!

File: src/DynamicNLPModels.jl [hunk 24]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+    LinearAlgebra.mul!(QAs0, Qf, As0[(ns * N + 1):end])
+
+    h0 += LinearAlgebra.dot(As0[(ns * N + 1):end], QAs0)
+
+    return (H = H, c = h, c0 = h0 / T(2))
+end
+
+
+
+function _build_H_blocks(Q, R, block_A::M, block_B::M, S::M, s0, N, Qf, K::MK) where {T, M <: AbstractMatrix, MK <: AbstractMatrix{T}}
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: genuine question: can we use a single `_build_H_blocks` function instead of two? Just to avoid code duplicate.

File: src/DynamicNLPModels.jl [hunk 25]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+    # quad term refers to the summation of Q, K^T RK, SK, and K^T S^T that is left and right multiplied by B in the Hessian
+    quad_term = similar(Q, ns, ns); fill!(quad_term, 0.0)
+
+    quad_term_B  = similar(block_B); fill!(quad_term_B, 0.0)
+    QfB = similar(block_B); fill!(QfB, 0.0)
+
+    quad_term_AB = similar(block_A, ns, nu) 
+    QfAB         = similar(block_A, ns, nu)
+
+    RK_STB    = similar(block_B, nu, nu)
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: also: rule of thumb, for an old-school guy: always allocate the arrays at the beginning of the function (in C, your compiler should tell you to do so)

Comment (frapac) [reply src/DynamicNLPModels.jl]: if we stick to that convention, it's a lot easier to assess how much memory is allocated inside one function

File: src/DynamicNLPModels.jl [hunk 26]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
-    return (J = G, lcon = vec(block_gl), ucon = vec(block_gu), As0 = As0)
+
+    EAs0 = similar(block_B, nc)
+    
+    for i in 1:N
+        if i != N
+            B_sub_block = block_B[(1 + (i - 1) * ns):(i * ns),:]
+
+            LinearAlgebra.mul!(EB, E_FK, B_sub_block)
+            G[(1 + nc * i):(nc * (i + 1)), :] = EB
```
Comment (frapac) [inline src/DynamicNLPModels.jl:1109]: broadcast!

File: src/DynamicNLPModels.jl [hunk 27]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
+end
+
+function _set_G_blocks!(G, dl, du, block_B, block_A, s0, E, F, K::MK, N) where {T, MK <: AbstractMatrix{T}}
+    ns = size(E, 2)
+    nu = size(F, 2)
+    nc = size(E, 1)
 
-    block_E_mul = similar(block_E)
-    LinearAlgebra.copyto!(block_E_mul, block_E)
+    G[1:nc, :] = F
```
Comment (frapac) [inline src/DynamicNLPModels.jl:1089]: broadcast!

File: src/DynamicNLPModels.jl [hunk 28]
```diff
@@ -1015,141 +982,308 @@ function _build_block_matrices(
...
-    LinearAlgebra.axpy!(-1, EAs0, block_gl)
-    LinearAlgebra.axpy!(-1, EAs0, block_gu)
-  
-    return (J = G, lcon = vec(block_gl), ucon = vec(block_gu), As0 = As0)
+
+    EAs0 = similar(block_B, nc)
+    
+    for i in 1:N
+        if i != N
+            B_sub_block = block_B[(1 + (i - 1) * ns):(i * ns),:]
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: please consider using a view here

**Review by frapac — CHANGES_REQUESTED** [review]
I think this PR is still a work in progress, but most requests are easy to address. 

As a rule of thumb, I would advise
1. Use the broadcast operator consistently and ensure we are not allocating any new array inside a for loop
2. Move all the allocations (`similar`) at the beginning of the functions, so we can quickly assess how much memory is allocated
3. Use the notations from Kerrigan's paper consistently

File: src/DynamicNLPModels.jl (hunk 7 above)
Comment (dlcole3) [reply src/DynamicNLPModels.jl]: Not quite. The `G` matrix makes up part of the dense Jacobian. For the sake of memory, I actually just save the first block column of the `G` matrix, and then I fill the Jacobian separately. The Jacobian is also made up of the variable bounds (which become algebraic constraints). I have a couple prepared equations I will show tomorrow.

File: src/DynamicNLPModels.jl (hunk 9 above)
Comment (dlcole3) [reply src/DynamicNLPModels.jl]: Woah, I wish I knew about the `repeat` function a long time ago :) I will be adding that in several places. Thank you!

File: src/DynamicNLPModels.jl (hunk 22 above)
Comment (dlcole3) [reply src/DynamicNLPModels.jl]: @frapac Yes, I believe so. I played with several things, but the `view` was the one that led to the smallest number of allocations. I honestly don't know why, but when I used other things (like `H[row_range, col_range]`), this was far less efficient and took lots more memory.

File: src/DynamicNLPModels.jl (hunk 24 above)
Comment (dlcole3) [reply src/DynamicNLPModels.jl]: I think we can. Previously, `_build_H_blocks` had to have more memory allocation within it to account for more multiplications. However, now the memory allocation difference is negligible under the new formulation. I checked the allocation for a large scale problem, and the two functions with and without `K` used 2.313 GiB vs. 2.311 GiB respectively.

File: src/DynamicNLPModels.jl (hunk 1 above)
Comment (dlcole3) [reply src/DynamicNLPModels.jl]: @frapac can I do `fill!(S, 0)` within the function argument? This is setting the default value. I used `.= 0` because I wasn't sure how it would manage a `fill!` function within the function arguments.

**Comment by dlcole3** [comment]
I also updated the `build_thinplate.jl` file in the `examples/` folder. This used the old `DenseLQDynamicBlocks` structure

**Comment by dlcole3** [comment]
@frapac I made one more change after requesting reviews again (sorry, I should've added it before requesting reviews). I realized I am allocating extra memory for the Jacobian. I essentially was buliding the Jacobian twice. I removed the extra `J1`, `J2`, and `J3` matrices in the most recent commit.

File: src/DynamicNLPModels.jl [hunk 29]
```diff
@@ -749,148 +704,148 @@ function _build_dense_lq_dynamic_model(dnlp::LQDynamicData{T,V,M,MK}) where {T,
...
     else
         for i in 1:N
-            row_range   = (1 + (i - 1) * num_real_bounds):(i * num_real_bounds)
-            J3[row_range, :] .= KBI[(1 + nu * (i - 1)):(nu * i), :][bool_vec_u, :]
-            KAs0_bounds[row_range]      .= KAs0[(1 + nu * (i - 1)):(nu * i)][bool_vec_u]
+            row_range              = (1 + (i - 1) * num_real_bounds_u):(i * num_real_bounds_u)
+            KAs0_bounds[row_range] = KAs0[(1 + nu * (i - 1)):(nu * i)][bool_vec_u]
+
+            for j in 1:(N - i +1)
+            J[(offset_u + 1 + (i + j - 2) * num_real_bounds_u):(offset_u + (i + j - 1) * num_real_bounds_u), (1 + nu * (i - 1)):(nu * i)] = @view(KBI[(1 + (j - 1) * nu):(j * nu), :][bool_vec_u, :])
```
Comment (frapac) [inline src/DynamicNLPModels.jl]: indentation!

**Review by frapac — APPROVED** [review]
Just an indentation issue, but otherwise it looks good to merge. 

Good work!
